### PR TITLE
Avoid compound literal struct init for better C/C++ compat.

### PR DESCRIPTION
**Pull request by ScottTodd**
Progress on this downstream build issue: https://github.com/nod-ai/shark-ai/issues/534.

I haven't figured out specifically why the downstream build hits errors here while upstream is fine, but these changes aren't that intrusive.

### CODE REVIEW

File: runtime/src/iree/hal/command_buffer.h [hunk 1]
```diff
@@ -115,13 +115,25 @@ typedef struct iree_hal_buffer_ref_t {
...
     iree_hal_buffer_t* buffer, iree_device_size_t offset,
     iree_device_size_t length) {
-  return (iree_hal_buffer_ref_t){0, 0, buffer, offset, length};
+  iree_hal_buffer_ref_t hal_buffer_ref;
+  hal_buffer_ref.reserved = 0;
+  hal_buffer_ref.buffer_slot = 0;
+  hal_buffer_ref.buffer = buffer;
+  hal_buffer_ref.offset = offset;
+  hal_buffer_ref.length = length;
+  return hal_buffer_ref;
```
Comment (benvanik) [inline runtime/src/iree/hal/command_buffer.h]:
```suggestion
  iree_hal_buffer_ref_t buffer_ref = {0};
  buffer_ref.reserved = 0;
  buffer_ref.buffer_slot = 0;
  buffer_ref.buffer = buffer;
  buffer_ref.offset = offset;
  buffer_ref.length = length;
  return buffer_ref;
```

File: runtime/src/iree/hal/command_buffer.h [hunk 2]
```diff
@@ -115,13 +115,25 @@ typedef struct iree_hal_buffer_ref_t {
...
     uint32_t buffer_slot, iree_device_size_t offset,
     iree_device_size_t length) {
-  return (iree_hal_buffer_ref_t){0, buffer_slot, NULL, offset, length};
+  iree_hal_buffer_ref_t hal_buffer_ref;
+  hal_buffer_ref.reserved = 0;
+  hal_buffer_ref.buffer_slot = 0;
+  hal_buffer_ref.buffer = NULL;
+  hal_buffer_ref.offset = offset;
+  hal_buffer_ref.length = length;
+  return hal_buffer_ref;
```
Comment (benvanik) [inline runtime/src/iree/hal/command_buffer.h]:
```suggestion
  iree_hal_buffer_ref_t buffer_ref = {0};
  buffer_ref.reserved = 0;
  buffer_ref.buffer_slot = 0;
  buffer_ref.buffer = NULL;
  buffer_ref.offset = offset;
  buffer_ref.length = length;
  return buffer_ref;
```

File: runtime/src/iree/base/alignment.h [hunk 3]
```diff
@@ -238,19 +238,19 @@ static inline iree_page_range_t iree_page_range_union(
     const iree_page_range_t a, const iree_page_range_t b) {
   iree_host_size_t start = iree_min(a.offset, b.offset);
   iree_host_size_t end = iree_max(a.offset + a.length, b.offset + b.length);
-  return (iree_page_range_t){
-      /*.offset=*/start,
-      /*.length=*/end - start,
-  };
+  iree_page_range_t page_range;
```
Comment (benvanik) [inline runtime/src/iree/base/alignment.h]:
```suggestion
  iree_page_range_t page_range = {0};
```

File: runtime/src/iree/base/alignment.h [hunk 4]
```diff
@@ -238,19 +238,19 @@ static inline iree_page_range_t iree_page_range_union(
...
 }
 
 // Aligns a byte range to page boundaries defined by |page_alignment|.
 static inline iree_page_range_t iree_align_byte_range_to_pages(
     const iree_byte_range_t byte_range, iree_host_size_t page_alignment) {
-  return (iree_page_range_t){
-      /*.offset=*/iree_host_align(byte_range.offset, page_alignment),
-      /*.length=*/iree_host_align(byte_range.length, page_alignment),
-  };
+  iree_page_range_t page_range;
```
Comment (benvanik) [inline runtime/src/iree/base/alignment.h]:
```suggestion
  iree_page_range_t page_range = {0};
```

**Review by benvanik — APPROVED** [review]

**Comment by ScottTodd** [comment]
Tests failing, will investigate + fix...

**Pull request merged**